### PR TITLE
[#647] Added 'DropzoneTrait' for multi-file drag-and-drop step.

### LINKED ISSUE #647: Add 'DropzoneTrait' with multi-file drag-and-drop step.
# Add `DropzoneTrait` with multi-file drag-and-drop step

## Summary

New trait that lets a scenario simulate a real drag-and-drop gesture, dropping one or more files onto a Dropzone-style target in a single native event:

```gherkin
When I drop the file "a.pdf" on the ".dropzone" dropzone
When I drop the following files on the ".dropzone" dropzone:
  | a.pdf |
  | b.pdf |
  | c.pdf |
```

## Motivation

Today there is no way to test multi-file drag-and-drop uploads from Behat.

Mink's `Element::attachFile()` writes to a hidden `<input type="file">` and runs each upload sequentially: file A is fully uploaded before file B starts. Real users select several files in their OS file picker, or drag a group of files from their desktop, and release them together — which fires a **single** `drop` event whose `dataTransfer.files` contains all of them. Upload pipelines react to that by kicking off concurrent uploads, and several common failure modes only show up on that path:

- Race in a per-file deduplication map (last write wins; one file gets silently dropped).
- Status indicator updated by whichever upload finishes last instead of aggregated state.
- Per-file error handler that assumes serial execution and clobbers the UI on the second concurrent failure.
- Server-side queue/lock that handles N=1 but deadlocks at N>1 within the same request window.

None of these reproduce under sequential `attachFile()` calls. They reproduce reliably under a single multi-file `drop`.

The underlying Dropzone.js library is widely deployed — in the Drupal `dropzonejs` contrib module, in parts of Drupal Media Library workflows, as the file widget in some Webform configurations, and in many non-Drupal apps. The trait is browser-level (it dispatches a DOM event), so it belongs in the generic namespace (`src/`), not under `src/Drupal/`.

## Proposed steps

| Step | Method |
|---|---|
| `When I drop the file :path on the :selector dropzone` | `dropzoneDropFile` |
| `When I drop the following files on the :selector dropzone:` (TableNode, one path per row) | `dropzoneDropFiles` |

Both use `#[When(...)]` tuple annotations. Paths are resolved against the Mink `files_path` parameter, consistent with `Mink::attachFileToField`.

## Implementation notes

The browser-level gesture this trait needs to reproduce:

1. Resolve the target element from the CSS selector. Fail with a clear exception if missing.
2. For each path:
   - Resolve the absolute path against `files_path`. Fail clearly if the fixture file is missing.
   - Create a hidden `<input type="file">` in the DOM with a unique id (the "holder").
   - Attach the fixture to that holder via `Element::attachFile()`, so the browser ingests the file via the normal upload pathway.
3. Build a single `DataTransfer`, copy `holder.files[0]` from each holder into it via `dt.items.add(...)`.
4. Dispatch `new DragEvent('drop', { bubbles: true, cancelable: true, dataTransfer: dt })` on the target.
5. Remove the holder inputs to keep the DOM clean.

All JavaScript can run via `Session::executeScript`. No external dependencies. `@javascript`-only (a headless browser session is required).

The single-file step is a thin wrapper that builds a one-row table and calls the multi-file method, so the gesture path is exercised identically in both cases.

## Scope boundaries

- The trait fires the gesture only. It does not assert anything about what Dropzone does next (file list, progress, error messages). Assertions on resulting UI belong to the consumer using existing element/text steps.
- Does not require the target to be a Dropzone.js instance specifically; any element that handles native `drop` events will work. The step is named after the user-facing concept ("dropzone"), not the library.
- No "drop folder" or directory-tree variant. Browsers expose folder drops via `dataTransfer.items` with `webkitGetAsEntry()`, which is a different surface and a separate request if needed.
- No retry/wait semantics. If the consumer needs to wait for asynchronous Dropzone state to settle, that is `WaitTrait` territory.

## Acceptance criteria

- [ ] New file `src/DropzoneTrait.php` in the generic (non-Drupal) namespace.
- [ ] Two methods with `#[When(...)]` tuple annotations as above.
- [ ] Fixture-path resolution via the Mink `files_path` parameter.
- [ ] Clear exceptions for missing target element and missing fixture files.
- [ ] Feature tests under `tests/behat/features/` covering single-file and multi-file drops, plus negative cases (missing target, missing fixture).
- [ ] `@trait:DropzoneTrait` subprocess tests.
- [ ] `ahoy update-docs` run.
- [ ] `ahoy lint` passes.

### PR DESCRIPTION
**Pull request by AlexSkrypnyk**
Closes #647

## Summary

Adds `DropzoneTrait` to the generic namespace, providing two `When` steps that simulate a real multi-file drag-and-drop gesture by dispatching a single native `drop` event whose `dataTransfer.files` contains all requested files simultaneously. Unlike Mink's `attachFile`, which writes files to a hidden `<input type="file">` one at a time (sequential uploads), this trait reproduces the concurrent-upload code path that real users trigger - exposing race conditions in dedup maps, status-indicator clobbering, queue deadlocks, and server-side handlers that only manifest under the multi-file path. Fixture paths are resolved against the Mink `files_path` parameter and validated to prevent path-traversal escapes.

## Changes

**New trait - `src/DropzoneTrait.php`**
- `When I drop the file :path on the :selector dropzone` - single-file convenience step (delegates to the multi-file step with a one-row TableNode)
- `When I drop the following files on the :selector dropzone:` - drops N files in a single native `drop` event; injects temporary off-screen `<input type="file">` elements to satisfy browser security, collects the `File` objects into a `DataTransfer`, dispatches `DragEvent('drop', ...)` on the resolved CSS target, then removes the holder inputs
- `dropzoneResolvePath()` - resolves each path against `files_path`, rejects missing files and paths that escape the configured base directory via `realpath` comparison

**New tests - `tests/behat/features/dropzone.feature`**
- 4 positive `@javascript @phpserver` scenarios: single-file drop, multi-file drop (verifies a single `drop` event fires), drop on a non-default selector, two consecutive drops in one scenario
- 3 negative `@trait:DropzoneTrait` scenarios: missing target element, missing single fixture file, missing file in a multi-file list

**New fixture - `tests/behat/fixtures/dropzone.html`**
- Minimal HTML page with a `.dropzone` primary target and a `#secondary-zone` secondary target; JavaScript listeners append dropped filenames to separate output lists and increment a shared event counter, letting scenarios assert both which files landed where and how many `drop` events fired

**Registration - `tests/behat/bootstrap/FeatureContext.php`**
- Imported and applied `DropzoneTrait` alongside the existing generic traits

**Docs - `STEPS.md`, `README.md`**
- Regenerated by `ahoy update-docs`; `DropzoneTrait` entry added to the trait table in both files

## Before / After

```
BEFORE (Mink attachFile - sequential, one event per file)
─────────────────────────────────────────────────────────
  attachFile("document.pdf")  →  upload A starts → upload A finishes
  attachFile("image.png")     →  upload B starts → upload B finishes
  attachFile("text.txt")      →  upload C starts → upload C finishes

  Event sequence: input:change, input:change, input:change
  Concurrent upload window: NONE (each file finishes before next starts)
  Race conditions reproduced: NO

AFTER (DropzoneTrait - all files in one drop event)
───────────────────────────────────────────────────
  dropzoneDropFiles(".dropzone", ["document.pdf", "image.png", "text.txt"])
    │
    ├── inject hidden <input id="holder_0"> → attachFile → File A
    ├── inject hidden <input id="holder_1"> → attachFile → File B
    ├── inject hidden <input id="holder_2"> → attachFile → File C
    │
    └── dt = new DataTransfer()
        dt.items.add(File A)
        dt.items.add(File B)
        dt.items.add(File C)
        target.dispatchEvent(new DragEvent('drop', { dataTransfer: dt }))
        (remove holder inputs)

  Event sequence: drop (dataTransfer.files.length === 3)
  Concurrent upload window: YES - all three files arrive together
  Race conditions reproduced: YES
```

### Overview
Added a new Behat trait that simulates real multi-file drag-and-drop gestures by firing a single native `drop` event with all files in a `DataTransfer`, enabling tests of concurrent-upload code paths that Mink's sequential `attachFile` cannot reproduce.

### Step Definitions
Two new `When` steps (compliant with CONTRIBUTING.md):
- `When I drop the file :path on the :selector dropzone` — single-file convenience wrapper
- `When I drop the following files on the :selector dropzone:` — multi-file step accepting a table with one fixture path per row

Placeholders, phrasing (`the following`), action-verb `When I drop`, tuple format, and method naming (trait-prefixed) conform to the repository's step rules. No CONTRIBUTING.md violations found.

### Implementation Details
- New trait: src/DropzoneTrait.php
  - dropzoneDropFile(string $path, string $selector): delegates to the multi-file step.
  - dropzoneDropFiles(string $selector, TableNode $paths): resolves fixture paths, injects temporary off-screen `<input type="file">` holders, uses Mink's attachFileToField to populate each holder, aggregates holder.files into a single `DataTransfer`, dispatches one browser-native `DragEvent('drop', { bubbles: true, cancelable: true, dataTransfer: dt })` on the resolved CSS target, and removes holders.
  - dropzoneResolvePath(string $path): trims and validates input, requires a configured Mink `files_path`, canonicalizes `files_path` with `realpath()` and validates it is a directory, constructs the full path, checks file existence, canonicalizes the fixture path and rejects paths outside the canonical `files_path` directory (prevents path traversal). Throws clear RuntimeExceptions for empty paths, missing/invalid `files_path`, missing fixtures, and out-of-bounds fixtures.
- All DOM/JS interactions run via Session::executeScript; temporary inputs are positioned off-screen (fixed left -9999px, opacity 0).
- Throws ElementNotFoundException when the CSS target is missing at resolution time and throws a JS Error if the target disappears between resolution and dispatch.

### Tests & Fixtures
- tests/behat/features/dropzone.feature: positive scenarios (single-file, multi-file verifying a single drop event, non-default selector, consecutive drops) and negative `@trait` scenarios (missing target, missing single fixture, missing file in multi-file list).
- tests/behat/fixtures/dropzone.html: two drop zones, event counter, and JS handlers that read `event.dataTransfer.files` and append filenames.
- tests/behat/bootstrap/FeatureContext.php: DropzoneTrait registered via use statement.

### Documentation & Misc
- README.md: added DropzoneTrait to the generic steps index.
- STEPS.md: added full DropzoneTrait documentation and examples.
- Docs regenerated via ahoy update-docs.

### Scope & Acceptance
- Trait only simulates the drop gesture; assertions about Dropzone behavior remain the consumer's responsibility.
- No folder-drop support, no retry/wait semantics.
- Acceptance criteria met: new src/DropzoneTrait.php with two #[When(...)] methods, robust fixture-path resolution and canonicalization, clear exceptions for missing targets/fixtures, feature tests (positive and negative), and documentation updates.

### CODE REVIEW

File: src/DropzoneTrait.php [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+    $base = rtrim((string) realpath((string) $files_path), DIRECTORY_SEPARATOR);
+    $full_path = $base . DIRECTORY_SEPARATOR . ltrim($path, '/');
+
+    if (!is_file($full_path)) {
+      throw new \RuntimeException(sprintf('The fixture file "%s" does not exist.', $full_path));
+    }
+
+    $resolved = realpath($full_path);
+    if ($resolved === FALSE || !str_starts_with($resolved, $base . DIRECTORY_SEPARATOR)) {
+      throw new \RuntimeException(sprintf('The fixture file "%s" is outside the configured "files_path".', $path));
```
[the remark this answers is not in the document]
Comment (AlexSkrypnyk) [reply src/DropzoneTrait.php]: Fixed in 247b414. Validated 'realpath(files_path)' upfront and threw a 'RuntimeException' if it returns 'FALSE' or is not a directory, so '$base' can no longer be empty and the containment check cannot be trivially satisfied by a leading slash. Also normalized 'ltrim($path, "/\\\\")' to strip both POSIX and Windows path separators.
